crates/librqbit/src/blocklist.rs

```diff
@@ -0,0 +1,273 @@
+use anyhow::{Context, Result};
+use async_compression::tokio::bufread::GzipDecoder;
+use futures::TryStreamExt;
+use intervaltree::IntervalTree;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::pin::Pin;
+use std::str::FromStr;
+use tokio::io::{AsyncBufRead, AsyncRead};
+use tokio::{io::AsyncBufReadExt, io::BufReader};
+use tokio_util::io::StreamReader;
+use tracing::{debug, info, trace};
+use url::Url;
+
+pub struct Blocklist {
+    // ipv4 and ipv6 do not overlap
+    // see: https://www.rfc-editor.org/rfc/rfc4291#section-2.5.5
+    blocked_ranges: IntervalTree<IpAddr, ()>,
+}
+
+impl Blocklist {
+    pub fn empty() -> Self {
+        return Self::new(std::iter::empty());
+    }
+
+    pub fn new(ip_ranges: impl IntoIterator<Item = std::ops::Range<IpAddr>>) -> Self {
+        Self {
+            blocked_ranges: IntervalTree::from_iter(ip_ranges.into_iter().map(|r| (r, ()))),
+        }
+    }
+
+    pub async fn load_from_url(url: &str) -> Result<Self> {
+        let parsed_url = Url::parse(url).context("Failed to parse URL")?;
+
+        if parsed_url.scheme() == "file" {
+            let path = parsed_url
+                .to_file_path()
+                .map_err(|_| anyhow::anyhow!("Failed to convert file URL to path"))?;
+            return Self::load_from_file(path.to_str().unwrap()).await;
```

Comment:
rather than using .unwrap() here it would be better to pass &Path (or AsRef<Path>) into load_from_file

In any case let's strive not use .unwrap() in librqbit unless it's 100% clear it will never fail